rir/src/interpreter/interp.cpp

```diff
@@ -491,6 +491,264 @@ RIR_INLINE SEXP legacyCall(const CallContext& call, Context* ctx) {
     return res;
 }
 
+SEXP tryFastSpecialCall(const CallContext& call, Context* ctx) {
+    SLOWASSERT(call.hasStackArgs() && !call.hasNames());
+    return nullptr;
+}
+
+SEXP tryFastBuiltinCall(const CallContext& call, Context* ctx) {
+    SLOWASSERT(call.hasStackArgs() && !call.hasNames());
+
+    static constexpr size_t MAXARGS = 32;
+    std::array<SEXP, MAXARGS> args;
+    auto nargs = call.suppliedArgs;
+
+    if (nargs > MAXARGS)
+        return nullptr;
+
+    for (size_t i = 0; i < call.suppliedArgs; ++i) {
+        auto arg = call.stackArg(i);
+        if (TYPEOF(arg) == PROMSXP)
+            arg = PRVALUE(arg);
+        if (arg == R_UnboundValue || arg == R_MissingArg ||
+            ATTRIB(arg) != R_NilValue)
+            return nullptr;
+        args[i] = arg;
+    }
+
+    switch (call.callee->u.primsxp.offset) {
+    case 90: { // c
+        if (nargs == 0)
+            return R_NilValue;
+
+        auto type = TYPEOF(args[0]);
+        if (type != REALSXP && type != LGLSXP && type != INTSXP)
+            return nullptr;
+        long total = XLENGTH(args[0]);
+        for (size_t i = 1; i < nargs; ++i) {
+            auto thistype = TYPEOF(args[i]);
+            if (thistype != REALSXP && thistype != LGLSXP && thistype != INTSXP)
+                return nullptr;
+
+            if (thistype == INTSXP && type == LGLSXP)
+                type = INTSXP;
+
+            if (thistype == REALSXP && type != REALSXP)
+                type = REALSXP;
+
+            total += XLENGTH(args[i]);
+        }
+
+        if (total == 0)
+            return nullptr;
+
+        long pos = 0;
+        auto res = Rf_allocVector(type, total);
+        for (size_t i = 0; i < nargs; ++i) {
+            auto len = XLENGTH(args[i]);
+            for (long j = 0; j < len; ++j) {
+                assert(pos < total);
+                if (type == REALSXP) {
+                    if (TYPEOF(args[i]) == REALSXP) {
+                        REAL(res)[pos++] = REAL(args[i])[j];
+                    } else {
+                        SLOWASSERT(NA_INTEGER == NA_LOGICAL);
```

Comment:
Maybe comment why this is important (here you're handling the integer and logical cases, and both are handled the same way).